### ISSUE TITLE
cant save keyboard configuration

### ISSUE THREAD

**Issue by axaluss** [mupen64plus/mupen64plus-ui-python#2]
When i change the key bindings in settings>plugins>input>configure they are not saved.
My changes are always reverted to default. 

How about a "save as..." button for the input config?

**Comment by deathpancake** [comment]
The same thing happened to me, I had to manually edit the .cfg file as well. Quite an oversight!

**Comment by Korvox** [comment]
So I'm having this issue as well, running master on Arch. I think I'm getting a bug of a different variety though, because in my case...

```
File "mupen64plus-ui-python/src/m64py/frontend/input.py", line 420, in get_sdl_key
    key = QKeySequence(text).__int__()
NameError: name 'QKeySequence' is not defined
``` 

Which is just an obvious missing

    from PyQt5.QtGui import QKeySequence

If I add that to the top though...

    key = QKeySequence(text).__int__()
    AttributeError: 'QKeySequence' object has no attribute '__int__'

Next line needs an int code for the key to pass to a converter to SDL2's keycodes, and never having touched pyqt before trying to find documentation on this was a lost cause for me. The general Qt methods don't provide a way to generate integral key codes from sequences in the primary docs, though.

I'm gonna assume that code is almost never invoked, and if I just pass --sdl2 to m64py I take the main codepath and can assign most keys again, but strangely assigning "Shift" isn't working while other modifiers like ctrl and alt are. I'm curious if its a syntax error somewhere. So I just printed out the text being passed into the get_sdl_key function and saw... oddities.

```
Retur&n
Retur&n
,
,
.
.
Le&ft
Le&ft
L&eft
L&eft
&Right
&Right
&Up
&Up
D&own
D&own
```

Looks like mnemonic codes to me, and they seem to be in the SDL keycodes? Some keys I can reassign and some I cannot. ctrl, space, and up work. Left and shift do not. But I digress here, maybe this info will help with this bug, I recognize the rabbit hole into learning way too much about pyqt and the sdl2 bindings for Python all weekend when I see it.

... I just wanted to rebind my keys and play some Banjo! ;x

**Comment by slimjim27** [comment]
hey just figured id ask um im needing help trying to get my usb controller to work on m64py its got a messed up joystick so i can only use one (the rightstick) so i use it for n64 emus specifically how do i get m64py to save my manual controller configuration?

**Comment by illuminati7** [comment]
Does anyone know where the config file would be stored on OS X El Capitan? I'm trying to help a friend but I never used OS X

**Comment by tormynator** [comment]
@illuminati7 According to this site: https://conchurnavid.github.io/docs/KeyboardSetup.html, the config file should be here:
~/.config/mupen64plus/mupen64plus.cfg

**Comment by soreau** [comment]
@MrBluewoods 
> I just installed and tried the latest release and I can confirm that it still occurs, but only for USB controller inputs. That is: It will save my settings if I assign keyboard keys to the N64 controller buttons, but not if I assign USB controller buttons to N64 controller buttons. Those will revert to their previous values as soon as I close the input configuration window.

I had this problem and managed to fix it with this:

diff --git a/src/m64py/frontend/input.py b/src/m64py/frontend/input.py
index f1765e6..b71e218 100644
--- a/src/m64py/frontend/input.py
+++ b/src/m64py/frontend/input.py
@@ -328,6 +328,7 @@ class Input(QDialog, Ui_InputDialog):
                     v = "key(%s)" % value
                     self.config.set_parameter(key, v.encode())
                 else:
+                    self.config.set_parameter(key, widget.text().encode())
                     continue
 
     def get_axis(self, axis):

**Comment by inukaze** [comment]
I can't change the default keyboard values . i am writring the follow to the config, and dont work : 

```
[Input-SDL-Control1]

# Mupen64Plus SDL Input Plugin config parameter version number.  Please don't change this version number.
version = 2,000000
# Controller configuration mode: 0=Fully Manual, 1=Auto with named SDL Device, 2=Fully automatic
mode = 0
# Specifies which joystick is bound to this controller: -1=No joystick, 0 or more= SDL Joystick number
device = -1
# SDL joystick name (or Keyboard)
name = "Keyboard"
# Specifies whether this controller is 'plugged in' to the simulated N64
plugged = True
# Specifies which type of expansion pak is in the controller: 1=None, 2=Mem pak, 5=Rumble pak
plugin = 2
# If True, then mouse buttons may be used with this controller
mouse = False
# Scaling factor for mouse movements.  For X, Y axes.
MouseSensitivity = "2,00,2,00"
# The minimum absolute value of the SDL analog joystick axis to move the N64 controller axis value from 0.  For X, Y axes.
AnalogDeadzone = "4096,4096"
# An absolute value of the SDL joystick axis >= AnalogPeak will saturate the N64 controller axis value (at 80).  For X, Y axes. For each axis, this must be greater than the corresponding AnalogDeadzone value
AnalogPeak = "32768,32768"
# Digital button configuration mappings
DPad R = "key(275)"
DPad L = "key(276)"
DPad D = "key(274)"
DPad U = "key(273)"
Start = "key(13)"
Z Trig = "key(104)"
B Button = "key(116)"
A Button = "key(103)"
C Button R = "key(108)"
C Button L = "key(106)"
C Button D = "key(107)"
C Button U = "key(105)"
R Trig = "key(101)"
L Trig = "key(113)"
Mempak switch = "key(44)"
Rumblepak switch = "key(46)"
# Analog axis configuration mappings
X Axis = "key(97,100)"
Y Axis = "key(119,115)"
```

I dont know why , i set like the above , but works like from below : 

```
A Button = "key(304)"
Z Trig = "key(122)"
B Button = "key(306)"
```

For example. someone can help me to fix it. i compile manually the source code from git

**Comment by richard42** [comment]
gen2brain, shouldn't this issue be closed?

inukaze, are you using the M64Py front-end application?

**Comment by inukaze** [comment]
Richard42 yes the python ui stuck the main window if you wanna see I can record it

**Comment by gen2brain** [comment]
I thought this is fixed and only new release was missing. I will test again, currently I don't have any joystick, I wil try it probably for weekend.

**Comment by inukaze** [comment]
The CLI UI don't save neither is complete ignored the keyboard
configuration on ~/.config/mupen64plus/mupen64plus.cfg

Just take the values from file InputAutoCfg.ini

Not from the user configuration

**Comment by gen2brain** [comment]
And what do you have in config for controller configuration mode?

# Controller configuration mode: 0=Fully Manual, 1=Auto with named SDL Device, 2=Fully automatic                                                                                             
mode = 0

**Comment by inukaze** [comment]
Ok i had the follow inside my file ~/.config/mupen64plus/mupen64plus :

[Input-SDL-Control1]

# Mupen64Plus SDL Input Plugin config parameter version number.  Please
don't change this version number.
version = 2
# Controller configuration mode: 0=Fully Manual, 1=Auto with named SDL
Device, 2=Fully automatic
mode = 0
# Specifies which joystick is bound to this controller: -1=No joystick, 0
or more= SDL Joystick number
device = -1
# SDL joystick name (or Keyboard)
name = "Keyboard"
# Specifies whether this controller is 'plugged in' to the simulated N64
plugged = True
# Specifies which type of expansion pak is in the controller: 1=None, 2=Mem
pak, 5=Rumble pak
plugin = 2
# If True, then mouse buttons may be used with this controller
mouse = False
# Scaling factor for mouse movements.  For X, Y axes.
MouseSensitivity = "2.00,2.00"
# The minimum absolute value of the SDL analog joystick axis to move the
N64 controller axis value from 0.  For X, Y axes.
AnalogDeadzone = "1710,1710"
# An absolute value of the SDL joystick axis >= AnalogPeak will saturate
the N64 controller axis value (at 80).  For X, Y axes. For each axis, this
must be greater than the corresponding AnalogDeadzone value
AnalogPeak = "15708,15708"
# Digital button configuration mappings
DPad R = "key(275)"
DPad L = "key(276)"
DPad D = "key(274)"
DPad U = "key(273)"
Start = "key(13)"
Z Trig = "key(104)"
B Button = "key(116)"
A Button = "key(103)"
C Button R = "key(108)"
C Button L = "key(106)"
C Button D = "key(107)"
C Button U = "key(105)"
R Trig = "key(101)"
L Trig = "key(113)"
Mempak switch = "key(44)"
Rumblepak switch = "key(46)"
X Axis = "key(97,100)"
Y Axis = "key(119,115)"


But the keys working are diferent, the mapping working is :

[Keyboard]
plugged = True
plugin = 2
mouse = False
AnalogDeadzone = "4096,4096"
AnalogPeak = "32768,32768"
DPad R = "key(100)"
DPad L = "key(97)"
DPad D = "key(115)"
DPad U = "key(119)""
Start = "key(13)"
Z Trig = "key(122)"
B Button = "key(306)"
A Button = "key(304)"
C Button R = "key(108)"
C Button L = "key(106)"
C Button D = "key(107)"
C Button U = "key(105)"
R Trig = "key(99)"
L Trig = "key(120)"
Mempak switch = "key(44)"
Rumblepak switch = "key(46)"
X Axis = "key(276,275)"
Y Axis = "key(273,274)"

And you can see is diferent configuration. i must edit the InputAutoCfg.ini
to put my configuration working i wait.

2016-10-10 9:52 GMT-04:00 Milan Nikolic <notifications@github.com>:

**Comment by rz5** [comment]
To whom it may concern:

Since mupen64plus lacks a GUI, I resorted to this project. I noticed this GUI had some issues with saving the config file and came here to reconfirm the bug.

From that point, I searched for other GUIs and found RetroArch. There is a libretro version of mupen64 and I got the hang of using it through RA. Plus, the extra libretro features got me to stick with it till this day (been about 2 years since I posted in here).
